### PR TITLE
HW03: Nikita Dukin

### PR DESCRIPTION
**Pull request by Fant1k34**
Я устав, но было интересно!
Дз-шка заняла примерно 5.5 часов

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -1,40 +1,151 @@
-module Main where 
+module Main where
 
 import Text.Printf (printf)
 import Control.Monad (unless)
 
-data Expr
+-- Type for unary operator
+data Operator1 = Sqrt | Neg
```
Comment (kajigor) [inline Main.hs:7]: I don't remember unary negation being in the task, but okay :)

File: Main.hs [hunk 2]
```diff
@@ -1,40 +1,151 @@
...
+  show Sqrt = "√"
+  show Neg = "-"
 
-instance Eq Expr where 
-  (==) = undefined 
+instance Eq Operator1 where
+  Sqrt == Sqrt = True
+  Neg == Neg = True
+  Neg == Sqrt = False
+  Sqrt == Neg = False
```
Comment (kajigor) [inline Main.hs:17]: You can use wildcards so that not to write every possible combination

File: Main.hs [hunk 3]
```diff
@@ -1,40 +1,151 @@
...
 
-data Error 
 
-instance Show Error where 
-  show = undefined
+-- Type for binary operator
+data Operator2 = Div | Mul | Plus | Min deriving (Eq)
```
Comment (kajigor) [inline Main.hs:21]: What about `^`?

File: Main.hs [hunk 4]
```diff
@@ -1,40 +1,151 @@
...
-  (==) = undefined 
+instance Show Operator2 where
+  show Div = "/"
+  show Mul = "*"
+  show Plus = "+"
+  show Min = "-"
+
+
+-- Type for expression
+data Expr = Arg Double | Marg Operator1 Expr | CE Expr Operator2 Expr
```
Comment (kajigor) [inline Main.hs:31]: :+1:

File: Main.hs [hunk 5]
```diff
@@ -1,40 +1,151 @@
...
+  Marg op1 value1 == Marg op2 value2 =
+   op1 == op2 && value1 == value2
+  CE expr11 op1 expr12 == CE expr21 op2 expr22 =
+   expr11 == expr21 && op1 == op2 && expr12 == expr22
+  Arg value1 == Marg op value2 = False
+  Marg op value2 == Arg value1 = False
+  Arg value1 == CE expr1 op expr2 = False
+  CE expr1 op expr2 == Arg value1 = False
+  Marg op1 value2 == CE expr1 op2 expr2 = False
+  CE expr1 op2 expr2 == Marg op1 value2 = False
```
Comment (kajigor) [inline Main.hs:51]: Also wildcards here.

File: Main.hs (hunk 2 above)
Comment (kajigor) [reply Main.hs:17]: And if the instance is standard, then it's better to derive it.

File: Main.hs [hunk 6]
```diff
@@ -1,40 +1,151 @@
...
+
+
+data Error = OutOfPossibleValuesError Operator1 Double | ZeroDivisionError Double Double
```
Comment (kajigor) [inline Main.hs:54]: ` ZeroDivisionError Double Double` -- is there more than one possible denominator when dividing by 0?)

File: Main.hs [hunk 7]
```diff
@@ -1,40 +1,151 @@
...
+      _ -> Right (defineOperationEvaluator op result1 result2)
+    Left exception -> Left exception
+  Left exception -> Left exception
 
-eval :: Expr -> Either Error Double 
-eval = undefined 
 
 cases :: [(Expr, Either Error Double)]
-cases = undefined 
+cases = [
```
Comment (kajigor) [inline Main.hs:101]: :+1:

**Review by kajigor — CHANGES_REQUESTED** [review]

File: Main.hs (hunk 6 above)
Comment (Fant1k34) [reply Main.hs:54]: That's the origin way of legacy code ;)
Fixed it, thanks!

File: Main.hs (hunk 5 above)
Comment (Fant1k34) [reply Main.hs:51]: I hope that I get it in correct way, so I used "_ == _ = False"
Thanks for comment, it made code MUCH better ;)

File: Main.hs (hunk 3 above)
Comment (Fant1k34) [reply Main.hs:21]: Forgot ;)
Thank you, I added this operator, and added a new Error (IncorrectDegreeOfValue)

File: Main.hs (hunk 2 above)
Comment (Fant1k34) [reply Main.hs:17]: Okey, I fixed it with deriving ;)
Thank you!

File: Main.hs (hunk 1 above)
Comment (Fant1k34) [reply Main.hs:7]: Because I wanted to support negative values too)

**Comment by kajigor** [comment]
Стало куда лучше :) 👍

**Pull request closed without merging**